task/dump-agent/src/udp.rs

```diff
@@ -0,0 +1,167 @@
+)>::MAX_SIZE;
+
+// We are receiving a (Header, Request) from the host
+const MAX_UDP_RX_SIZE: usize =
+    <(humpty::udp::Header, humpty::udp::Request)>::MAX_SIZE;
+
+// Check against packet sizes in the TOML file
+static_assertions::const_assert!(MAX_UDP_TX_SIZE <= 1024);
+static_assertions::const_assert!(MAX_UDP_RX_SIZE <= 1024);
+
+/// Grabs references to the static descriptor/buffer receive rings. Can only be
+/// called once.
+pub fn claim_statics() -> (
+    &'static mut [u8; MAX_UDP_RX_SIZE],
+    &'static mut [u8; MAX_UDP_TX_SIZE],
+) {
+    mutable_statics::mutable_statics! {
+        static mut TX_BUF: [u8; MAX_UDP_RX_SIZE] = [|| 0u8; _];
+        static mut RX_BUF: [u8; MAX_UDP_TX_SIZE] = [|| 0u8; _];
+    }
+}
+
+impl ServerImpl {
+    pub fn check_net(
+        &mut self,
+        rx_data_buf: &mut [u8],
+        tx_data_buf: &mut [u8],
+    ) {
+        match self.net.recv_packet(
+            SocketName::dump_agent,
+            LargePayloadBehavior::Discard,
+            rx_data_buf,
+        ) {
+            Ok(meta) => self.handle_packet(meta, rx_data_buf, tx_data_buf),
+            Err(RecvError::QueueEmpty | RecvError::ServerRestarted) => {
+                // Our incoming queue is empty or `net` restarted. Wait for more
+                // packets in dispatch_n, back in the main loop.
+            }
+            Err(RecvError::NotYours | RecvError::Other) => panic!(),
+        }
+    }
+
+    fn handle_packet(
+        &mut self,
+        mut meta: UdpMetadata,
+        rx_data_buf: &[u8],
+        tx_data_buf: &mut [u8],
+    ) {
+        let out_len =
+            match hubpack::deserialize(&rx_data_buf[0..meta.size as usize]) {
+                Ok((mut header, msg)) => {
+                    let r = self.handle_message(header, msg);
+                    header.version = humpty::udp::version::CURRENT;
+                    let reply = (header, r);
+                    Some(hubpack::serialize(tx_data_buf, &reply).unwrap())
+                }
+                Err(e) => {
+                    // We couldn't deserialize the header; give up
+                    ringbuf_entry!(Trace::DeserializeHeaderError(e));
+                    None
+                }
+            };
+
+        if let Some(out_len) = out_len {
+            meta.size = out_len as u32;
+            if let Err(e) = self.net.send_packet(
+                SocketName::dump_agent,
+                meta,
+                &tx_data_buf[..meta.size as usize],
+            ) {
+                // We'll drop packets if the outgoing queue is full;
+                // the host is responsible for retrying.
+                //
+                // Other errors are unexpected and panic.
+                //
+                // This includes ServerRestarted, because the server should only
+                // restart if the watchdog times out, and the watchdog should
+                // not be timing out, because we're literally replying to a
+                // packet here.
```

Comment:
I'm not completely sold on panicking on `ServerRestarted`. The server should only restart if the watchdog times out _today_, but this seems to bake in knowledge about the net server that could change in the future. If `net` panics, we also panic, which doesn't seem ideal. Is there a downside to moving `ServerRestarted` into the "just drop the packet" arm with `QueueFull`?

Reply:
This was copy-pasted from `transceivers`, but I'm not convinced – I've moved `ServerRestarted` into the drop-the-packet branch and updated the comment in 37990a26c834b996bfa8a203fda959a66ff5c26f